File: app/src/main/java/com/example/student_grade/StudentDetailActivity.java
package com.example.student_grade;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import com.example.student_grade.database.AppDatabase;
import com.example.student_grade.database.Converters;
import com.example.student_grade.database.Grade;
import com.example.student_grade.database.Student;
import com.example.student_grade.database.StudentRepository;
import com.example.student_grade.utils.ScoreColorHelper;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import java.util.ArrayList;
import java.util.List;

public class StudentDetailActivity extends AppCompatActivity {
    
    private FloatingActionButton btnBack;
    private TextView tvStudentName;
    private TableLayout tableGrades;
    private TextView tvAverage;
    private FloatingActionButton btnAddGrade;
    private MaterialButton btnDeleteStudent;
    private StudentRepository repository;
    private final Converters converters = new Converters();
    
    private Student currentStudent;
    private int studentId = -1;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_student_detail);
        
        // Get student ID from intent
        studentId = getIntent().getIntExtra("STUDENT_ID", -1);
        if (studentId == -1) {
            Toast.makeText(this, R.string.toast_student_not_found, Toast.LENGTH_SHORT).show();
            finish();
            return;
        }
        
        // Initialize views
        btnBack = findViewById(R.id.btnBack);
        tvStudentName = findViewById(R.id.tvStudentName);
        tableGrades = findViewById(R.id.tableGrades);
        tvAverage = findViewById(R.id.tvAverage);
        btnAddGrade = findViewById(R.id.btnAddGrade);
        btnDeleteStudent = findViewById(R.id.btnDeleteStudent);
        
        // Initialize database and repository
        AppDatabase database = AppDatabase.getDatabase(getApplicationContext());
        repository = new StudentRepository(database.studentDao());
        
        // Setup button click listeners
        btnBack.setOnClickListener(v -> finish());
        btnAddGrade.setOnClickListener(v -> showAddGradeDialog());
        btnDeleteStudent.setOnClickListener(v -> showDeleteConfirmationDialog());
        
        loadStudentDetails();
    }
    
    private void loadStudentDetails() {
        repository.getStudentById(studentId).observe(this, student -> {
            if (student == null) {
                Toast.makeText(
                    StudentDetailActivity.this,
                    R.string.toast_student_not_found,
                    Toast.LENGTH_SHORT
                ).show();
                finish();
                return;
            }
            
            currentStudent = student;
            displayStudentDetails(student);
        });
    }
    
    private void displayStudentDetails(Student student) {
        // Display student name
        tvStudentName.setText(student.getName());
        
        // Get grades list
        List<Grade> grades = converters.toGradeList(student.getGradesCsv());
        
        // Clear existing table rows
        tableGrades.removeAllViews();
        
        // Add grade rows with cards
        for (int index = 0; index < grades.size(); index++) {
            Grade grade = grades.get(index);
            
            // Create a card for each grade
            View cardView = LayoutInflater.from(this)
                .inflate(R.layout.item_grade_display, tableGrades, false);
            
            TextView tvSubject = cardView.findViewById(R.id.tvSubject);
            TextView tvScore = cardView.findViewById(R.id.tvScore);
            View scoreContainer = cardView.findViewById(R.id.scoreContainer);
            
            tvSubject.setText(grade.getSubject());
            tvScore.setText(String.valueOf(grade.getScore()));
            
            // Apply score-based gradient to score container
            android.graphics.drawable.Drawable gradient = 
                ScoreColorHelper.getGradientForScore(this, (double) grade.getScore());
            scoreContainer.setBackground(gradient);
            
            TableRow row = new TableRow(this);
            TableLayout.LayoutParams params = new TableLayout.LayoutParams(
                TableLayout.LayoutParams.MATCH_PARENT,
                TableLayout.LayoutParams.WRAP_CONTENT
            );
            row.setLayoutParams(params);
            row.addView(cardView);
            tableGrades.addView(row);
        }
        
        // Calculate and display average
        double average = 0.0;
        if (!grades.isEmpty()) {
            average = grades.stream()
                .mapToInt(Grade::getScore)
                .average()
                .orElse(0.0);
        }
        
        tvAverage.setText(getString(R.string.average_format, average));
    }
    
    private void showAddGradeDialog() {
        View dialogView = LayoutInflater.from(this).inflate(R.layout.dialog_add_grade, null);
        EditText etSubject = dialogView.findViewById(R.id.etDialogSubject);
        EditText etScore = dialogView.findViewById(R.id.etDialogScore);
        
        AlertDialog dialog = new AlertDialog.Builder(this)
            .setTitle(R.string.dialog_add_grade_title)
            .setView(dialogView)
            .setPositiveButton(R.string.add, (dialog1, which) -> {
                String subject = etSubject.getText() != null ? 
                    etSubject.getText().toString().trim() : "";
                String scoreText = etScore.getText() != null ? 
                    etScore.getText().toString().trim() : "";
                
                try {
                    int score = Integer.parseInt(scoreText);
                    
                    if (subject.isEmpty()) {
                        Toast.makeText(
                            this,
                            R.string.toast_subject_empty,
                            Toast.LENGTH_SHORT
                        ).show();
                    } else if (score < 0 || score > 100) {
                        Toast.makeText(
                            this,
                            R.string.toast_grade_invalid,
                            Toast.LENGTH_SHORT
                        ).show();
                    } else {
                        addGrade(subject, score);
                    }
                } catch (NumberFormatException e) {
                    Toast.makeText(
                        this,
                        R.string.toast_grade_invalid,
                        Toast.LENGTH_SHORT
                    ).show();
                }
            })
            .setNegativeButton(R.string.cancel, null)
            .create();
        
        dialog.show();
    }
    
    private void addGrade(String subject, int score) {
        if (currentStudent == null) return;
        
        List<Grade> currentGrades = new ArrayList<>(converters.toGradeList(currentStudent.getGradesCsv()));
        currentGrades.add(new Grade(subject, score));
        
        String updatedGradesJson = converters.fromGradeList(currentGrades);
        Student updatedStudent = new Student();
        updatedStudent.setId(currentStudent.getId());
        updatedStudent.setName(currentStudent.getName());
        updatedStudent.setGradesCsv(updatedGradesJson);
        
        repository.updateStudent(updatedStudent, new StudentRepository.RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                Toast.makeText(
                    StudentDetailActivity.this,
                    R.string.toast_grade_added,
                    Toast.LENGTH_SHORT
                ).show();
                
                loadStudentDetails();
            }
            
            @Override
            public void onError(Exception error) {
                Toast.makeText(
                    StudentDetailActivity.this,
                    "Error adding grade: " + error.getMessage(),
                    Toast.LENGTH_SHORT
                ).show();
            }
        });
    }
    
    private void showDeleteConfirmationDialog() {
        new AlertDialog.Builder(this)
            .setTitle(R.string.dialog_delete_title)
            .setMessage(R.string.dialog_delete_message)
            .setPositiveButton(R.string.delete, (dialog, which) -> deleteStudent())
            .setNegativeButton(R.string.cancel, null)
            .create()
            .show();
    }
    
    private void deleteStudent() {
        if (currentStudent == null) return;
        
        repository.deleteStudent(currentStudent, new StudentRepository.RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                Toast.makeText(
                    StudentDetailActivity.this,
                    R.string.toast_student_deleted,
                    Toast.LENGTH_SHORT
                ).show();
                
                finish();
            }
            
            @Override
            public void onError(Exception error) {
                Toast.makeText(
                    StudentDetailActivity.this,
                    "Error deleting student: " + error.getMessage(),
                    Toast.LENGTH_SHORT
                ).show();
            }
        });
    }
}

File: app/src/main/java/com/example/student_grade/StudentDetailActivity.kt
package com.example.student_grade

import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import com.example.student_grade.database.AppDatabase
import com.example.student_grade.database.Converters
import com.example.student_grade.database.Student
import com.example.student_grade.database.StudentRepository
import com.example.student_grade.utils.ScoreColorHelper
import com.google.android.material.button.MaterialButton
import com.google.android.material.floatingactionbutton.FloatingActionButton
import kotlinx.coroutines.launch

class StudentDetailActivity : AppCompatActivity() {
    
    private lateinit var btnBack: FloatingActionButton
    private lateinit var tvStudentName: TextView
    private lateinit var tableGrades: TableLayout
    private lateinit var tvAverage: TextView
    private lateinit var btnAddGrade: FloatingActionButton
    private lateinit var btnDeleteStudent: MaterialButton
    private lateinit var repository: StudentRepository
    private val converters = Converters()
    
    private var currentStudent: Student? = null
    private var studentId: Int = -1
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_student_detail)
        
        // Get student ID from intent
        studentId = intent.getIntExtra("STUDENT_ID", -1)
        if (studentId == -1) {
            Toast.makeText(this, R.string.toast_student_not_found, Toast.LENGTH_SHORT).show()
            finish()
            return
        }
        
        // Initialize views
        btnBack = findViewById(R.id.btnBack)
        tvStudentName = findViewById(R.id.tvStudentName)
        tableGrades = findViewById(R.id.tableGrades)
        tvAverage = findViewById(R.id.tvAverage)
        btnAddGrade = findViewById(R.id.btnAddGrade)
        btnDeleteStudent = findViewById(R.id.btnDeleteStudent)
        
        // Initialize database and repository
        val database = AppDatabase.getDatabase(applicationContext)
        repository = StudentRepository(database.studentDao())
        
        // Setup button click listeners
        btnBack.setOnClickListener {
            finish()
        }
        
        btnAddGrade.setOnClickListener {
            showAddGradeDialog()
        }
        
        btnDeleteStudent.setOnClickListener {
            showDeleteConfirmationDialog()
        }
        
        loadStudentDetails()
    }
    
    private fun loadStudentDetails() {
        lifecycleScope.launch {
            val student = repository.getStudentById(studentId)
            
            if (student == null) {
                Toast.makeText(
                    this@StudentDetailActivity,
                    R.string.toast_student_not_found,
                    Toast.LENGTH_SHORT
                ).show()
                finish()
                return@launch
            }
            
            currentStudent = student
            displayStudentDetails(student)
        }
    }
    
    private fun displayStudentDetails(student: Student) {
        // Display student name
        tvStudentName.text = student.name
        
        // Get grades list
        val grades = converters.toGradeList(student.gradesCsv)
        
        // Clear existing table rows
        tableGrades.removeAllViews()
        
        // Add grade rows with cards
        grades.forEachIndexed { index, grade ->
            // Create a card for each grade
            val cardView = layoutInflater.inflate(R.layout.item_grade_display, tableGrades, false)
            
            val tvSubject = cardView.findViewById<TextView>(R.id.tvSubject)
            val tvScore = cardView.findViewById<TextView>(R.id.tvScore)
            val scoreContainer = cardView.findViewById<View>(R.id.scoreContainer)
            
            tvSubject.text = grade.subject
            tvScore.text = grade.score.toString()
            
            // Apply score-based gradient to score container
            val gradient = ScoreColorHelper.getGradientForScore(this, grade.score.toDouble())
            scoreContainer.background = gradient
            
            val row = TableRow(this)
            row.layoutParams = TableLayout.LayoutParams(
                TableLayout.LayoutParams.MATCH_PARENT,
                TableLayout.LayoutParams.WRAP_CONTENT
            )
            row.addView(cardView)
            tableGrades.addView(row)
        }
        
        // Calculate and display average
        val average = if (grades.isNotEmpty()) {
            grades.map { it.score }.average()
        } else {
            0.0
        }
        
        tvAverage.text = getString(R.string.average_format, average)
    }
    
    private fun showAddGradeDialog() {
        val dialogView = layoutInflater.inflate(R.layout.dialog_add_grade, null)
        val etSubject = dialogView.findViewById<EditText>(R.id.etDialogSubject)
        val etScore = dialogView.findViewById<EditText>(R.id.etDialogScore)
        
        val dialog = AlertDialog.Builder(this)
            .setTitle(R.string.dialog_add_grade_title)
            .setView(dialogView)
            .setPositiveButton(R.string.add) { _, _ ->
                val subject = etSubject.text.toString().trim()
                val scoreText = etScore.text.toString().trim()
                val score = scoreText.toIntOrNull()
                
                if (subject.isEmpty()) {
                    Toast.makeText(
                        this,
                        R.string.toast_subject_empty,
                        Toast.LENGTH_SHORT
                    ).show()
                } else if (score == null || score < 0 || score > 100) {
                    Toast.makeText(
                        this,
                        R.string.toast_grade_invalid,
                        Toast.LENGTH_SHORT
                    ).show()
                } else {
                    addGrade(subject, score)
                }
            }
            .setNegativeButton(R.string.cancel, null)
            .create()
        
        dialog.show()
    }
    
    private fun addGrade(subject: String, score: Int) {
        val student = currentStudent ?: return
        
        val currentGrades = converters.toGradeList(student.gradesCsv).toMutableList()
        currentGrades.add(com.example.student_grade.database.Grade(subject, score))
        
        val updatedGradesJson = converters.fromGradeList(currentGrades)
        val updatedStudent = student.copy(gradesCsv = updatedGradesJson)
        
        lifecycleScope.launch {
            repository.updateStudent(updatedStudent)
            
            Toast.makeText(
                this@StudentDetailActivity,
                R.string.toast_grade_added,
                Toast.LENGTH_SHORT
            ).show()
            
            loadStudentDetails()
        }
    }
    
    private fun showDeleteConfirmationDialog() {
        AlertDialog.Builder(this)
            .setTitle(R.string.dialog_delete_title)
            .setMessage(R.string.dialog_delete_message)
            .setPositiveButton(R.string.delete) { _, _ ->
                deleteStudent()
            }
            .setNegativeButton(R.string.cancel, null)
            .create()
            .show()
    }
    
    private fun deleteStudent() {
        val student = currentStudent ?: return
        
        lifecycleScope.launch {
            repository.deleteStudent(student)
            
            Toast.makeText(
                this@StudentDetailActivity,
                R.string.toast_student_deleted,
                Toast.LENGTH_SHORT
            ).show()
            
            finish()
        }
    }
}

